### DISCUSSION TITLE
Support for Starrocks

### DISCUSSION THREAD

**Discussion by maver1ck** [sqlalchemy/alembic#1562] [Usage Questions]
Hi,
I'm working on support for Starrocks dialect in Alembic.

The problem is that Starrocks support updates only on non-primary key columns in table that has different primary key.

As a result I want to create `alembic_version` table with following structure:
```
CREATE TABLE `alembic_version` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT COMMENT "",
  `version_num` varchar(32) NULL COMMENT ""
) ENGINE=OLAP
PRIMARY KEY(`id`)
DISTRIBUTED BY HASH(`id`);
```

I'd like to create similar patch to Clickhouse dialect. 
Source:
https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/master/clickhouse_sqlalchemy/alembic/dialect.py#L39

My version:
```
def patch_alembic_version(context, **kwargs):
    migration_context = context._proxy._migration_context
    old_version = migration_context._version
    metadata = old_version.metadata
    metadata.remove(old_version)

    version = Table(
        "alembic_version",
        metadata,
        Column("id", types.BIGINT, autoincrement=True, primary_key=True),
        Column("version_num", types.VARCHAR(32), primary_key=False),
        starrocks_primary_key="id",
    )
    migration_context._version = version
```

The main problem is that alembic assumes that first column in `alembic_version` table is `version_num`.

As a result of this analysis I created PR to alembic repo that will select only `version_num` column even if `alembic_version` table has more than one column.
https://github.com/sqlalchemy/alembic/pull/1561
Unfortunally it was discarded.

How to proceed with this problem ?

**Comment by maver1ck** [comment]
Generally speaking Alembic should have plugable way of defining `alembic_version` table.
Currently multiple dialects (like clickhouse) are just patching private fields which should be avoided.

**Reply by zzzeek** [reply]
OK so see, this is a more generalized problem first of all, it doesnt make that much sense to have a PR that supports a required monkeypatch elsewhere, you'd just create the whole feature :)   

Here's some ideas:

1. can starrocks support a table that has *no* primary key?   set [version_table_pk](https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.version_table_pk) to False.  now the table would have just the one column and no primary key.  we can update these docs to include this use case.
2. If starrocks supported this, you could just put the "id" column second.    but maybe it doesnt support PK cols at the end of the table either.

let me know if starrocks can use item 1 above if that's the only issue.  

The feature I would propose is another environment parameter ``version_table_autoinc_pk``, which would add a surrogate integer primary key to the table and support that mode of use, including that the select() use would target the version_num column only.   it would use `Integer`, not `BIGINT`.   I dont think a database is going to have 2B migrations :)         We only merge PRs that have tests also so this would need a bunch of tests, we can help you with those

**Reply by maver1ck** [reply]
Hi @zzzeek 
Thank you for answering.
Regarding your ideas.

1. I tested this earlier. In Starrocks you can't use UPDATE and DELETE statements if table doesn't have PRIMARY KEY.
Reference: https://docs.starrocks.io/docs/sql-reference/sql-statements/table_bucket_part_index/UPDATE/
> Updates rows in a Primary Key table.

2. I tried this solution also. Starrocks requires primary column to be the first.
Reference: https://docs.starrocks.io/docs/table_design/table_types/primary_key_table/
> In the CREATE TABLE statement, the primary key columns must be defined before other columns.

I think we should proceed differently.
Definition of version table should be part of `DefaultImpl`

As a result I will be able to overwrite it in `StarrocksImpl`.

Then we can change this code 
```
        self._version = Table(
            version_table,
            MetaData(),
            Column("version_num", String(32), nullable=False),
            schema=version_table_schema,
        )
        if opts.get("version_table_pk", True):
            self._version.append_constraint(
                PrimaryKeyConstraint(
                    "version_num", name="%s_pkc" % version_table
                )
            )

        self.impl = ddl.DefaultImpl.get_by_dialect(dialect)(
            dialect,
            self.connection,
            self.as_sql,
            transactional_ddl,
            self.output_buffer,
            opts,
        )
        
```
        
To something like
```
        self.impl = ddl.DefaultImpl.get_by_dialect(dialect)(
            dialect,
            self.connection,
            self.as_sql,
            transactional_ddl,
            self.output_buffer,
            opts,
        )

       self._version = self.impl.get_version_table(opts)
```

What do you think ?

Of course I can prepare PR.

**Reply by zzzeek** [reply]
we already have options to affect how the version_table is constructed, so let's use my proposal for a new parameter ``version_table_autoinc_pk=True``  

@CaselIT any opinions?

**Reply by maver1ck** [reply]
@zzzeek I understand. Eventually this option should be removed probably.

And your solution doesn't handle Clickhouse version of this table. Please take a look here:
https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/master/clickhouse_sqlalchemy/alembic/dialect.py#L39

**Reply by zzzeek** [reply]
they'd have to change that part.

I dont understand why clickhouse hardcoded this monkeypatch when it doesnt actually work with how alembic is querying the table currently

**Reply by CaselIT** [reply]
> I dont understand why clickhouse hardcoded this monkeypatch when it doesnt actually work with how alembic is querying the table currently

I think it does work, since the date time they add is added as the second column.




> we already have options to affect how the version_table is constructed, so let's use my proposal for a new parameter `version_table_autoinc_pk=True`
> 
> @CaselIT any opinions?

not sure. maybe we could provide a way of passing a custom table into the configure function?
This would be validated to have a version_num column that's selectable. Any select would then need to only use this version_num column.

This would support any kind of create table.

**Reply by maver1ck** [reply]
@CaselIT In the meantime any chance to merge my original PR ? https://github.com/sqlalchemy/alembic/pull/1561
It will solve my issue before making final solution.

**Reply by zzzeek** [reply]
> @CaselIT In the meantime any chance to merge my original PR ? #1561 It will solve my issue before making final solution.

no, sorry, that PR is not suitable, that's not how we would do that and it only makes sense in terms of an arbitrary monkeypatch which makes no sense.

**Reply by zzzeek** [reply]
> > I dont understand why clickhouse hardcoded this monkeypatch when it doesnt actually work with how alembic is querying the table currently
> 
> I think it does work, since the date time they add is added as the second column.

oh, I didnt realize we're talking about two different dialects here, OK

> 
> > we already have options to affect how the version_table is constructed, so let's use my proposal for a new parameter `version_table_autoinc_pk=True`
> > @CaselIT any opinions?
> 
> not sure. maybe we could provide a way of passing a custom table into the configure function? This would be validated to have a version_num column that's selectable. Any select would then need to only use this version_num column.

I dont want that approach.  We then have to deal with these well-known and long-used parameters being redundant: `version_table`, `version_table_schema`, `version_table_pk`.  There is no reason the end-user should be changing the structure of the alembic_version table arbitrarily.  At best, this is a function of the impl, in this case the starrocks impl, however, as it stands the only issue is the PK constraint which can be removed.



> 
> This would support any kind of create table.

**Reply by zzzeek** [reply]
@maver1ck  please start with this patch

```diff
diff --git a/alembic/ddl/impl.py b/alembic/ddl/impl.py
index 2574688..4618024 100644
--- a/alembic/ddl/impl.py
+++ b/alembic/ddl/impl.py
@@ -21,7 +21,12 @@ from typing import TYPE_CHECKING
 from typing import Union
 
 from sqlalchemy import cast
+from sqlalchemy import Column
+from sqlalchemy import MetaData
+from sqlalchemy import PrimaryKeyConstraint
 from sqlalchemy import schema
+from sqlalchemy import String
+from sqlalchemy import Table
 from sqlalchemy import text
 
 from . import _autogen
@@ -136,6 +141,30 @@ class DefaultImpl(metaclass=ImplMeta):
         self.output_buffer.write(text + "\n\n")
         self.output_buffer.flush()
 
+    def version_table_impl(self, version_table, version_table_schema, version_table_pk):
+        """create the Table object for the version_table.
+
+        Provided as part of impl so that third party dialects can override
+        this.
+
+        .. versionadded:: 1.13.4
+
+        """
+        vt = Table(
+            version_table,
+            MetaData(),
+            Column("version_num", String(32), nullable=False),
+            schema=version_table_schema,
+        )
+        if version_table_pk:
+            vt.append_constraint(
+                PrimaryKeyConstraint(
+                    "version_num", name=f"{version_table}_pkc"
+                )
+            )
+
+        return vt
+
     def requires_recreate_in_batch(
         self, batch_op: BatchOperationsImpl
     ) -> bool:
diff --git a/alembic/runtime/migration.py b/alembic/runtime/migration.py
index 6cfe5e2..f652808 100644
--- a/alembic/runtime/migration.py
+++ b/alembic/runtime/migration.py
@@ -24,10 +24,7 @@ from typing import Union
 
 from sqlalchemy import Column
 from sqlalchemy import literal_column
-from sqlalchemy import MetaData
-from sqlalchemy import PrimaryKeyConstraint
-from sqlalchemy import String
-from sqlalchemy import Table
+from sqlalchemy import select
 from sqlalchemy.engine import Engine
 from sqlalchemy.engine import url as sqla_url
 from sqlalchemy.engine.strategies import MockEngineStrategy
@@ -190,18 +187,6 @@ class MigrationContext:
         self.version_table_schema = version_table_schema = opts.get(
             "version_table_schema", None
         )
-        self._version = Table(
-            version_table,
-            MetaData(),
-            Column("version_num", String(32), nullable=False),
-            schema=version_table_schema,
-        )
-        if opts.get("version_table_pk", True):
-            self._version.append_constraint(
-                PrimaryKeyConstraint(
-                    "version_num", name="%s_pkc" % version_table
-                )
-            )
 
         self._start_from_rev: Optional[str] = opts.get("starting_rev")
         self.impl = ddl.DefaultImpl.get_by_dialect(dialect)(
@@ -212,6 +197,12 @@ class MigrationContext:
             self.output_buffer,
             opts,
         )
+
+        self._version = self.impl.version_table_impl(
+            version_table, version_table_schema,
+            opts.get("version_table_pk", True)
+        )
+
         log.info("Context impl %s.", self.impl.__class__.__name__)
         if self.as_sql:
             log.info("Generating static SQL")
@@ -540,7 +531,7 @@ class MigrationContext:
                 return ()
         assert self.connection is not None
         return tuple(
-            row[0] for row in self.connection.execute(self._version.select())
+            row[0] for row in self.connection.execute(select(self._version.c.version_num))
         )
 
     def _ensure_version_table(self, purge: bool = False) -> None:

```


tests for this need to be in tests/test_version_table.py.  use a custom impl class which overrides the method.  we can help you write the tests if you dont know how

**Reply by CaselIT** [reply]
ok, that use case is likely better. That should cover also the clickhouse monkey patch. I wonder why they didn't ask for this feature upstream...

**Reply by CaselIT** [reply]
Hi @xzkostyan sorry for the ping. Would this patch https://github.com/sqlalchemy/alembic/discussions/1562#discussioncomment-11103520 work for you to avoid monkey patching the version table as done here? https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/f52087844b86a9524c2295185ac29274b293d129/clickhouse_sqlalchemy/alembic/dialect.py#L39-L58

**Reply by maver1ck** [reply]
@zzzeek 
Thanks. I will prepare PR tomorrow.

**Reply by maver1ck** [reply]
Please find my PR here:
https://github.com/sqlalchemy/alembic/pull/1563
